### PR TITLE
feat(searchbar): add search bar ui to home page

### PR DESCRIPTION
**Pull request by atticusofsparta**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,12 +28,16 @@
     "@types/react": "^18.0.17",
     "@types/react-dom": "^18.0.6",
     "@vitejs/plugin-react": "^2.1.0",
-    "babel-jest": "^29.2.2",
     "jest": "^29.2.2",
     "jest-environment-jsdom": "^29.2.2",
+    "jest-svg-transformer": "^1.0.0",
```
Comment (dtfiedler) [inline package.json:33]: this isn't doing anything, remove

File: package.json [hunk 2]
```diff
@@ -28,12 +28,16 @@
...
     "jest-transform-stub": "^2.0.0",
     "prettier": "^2.7.1",
     "typescript": "^4.6.4",
-    "vite": "^3.1.0"
+    "vite": "^3.1.0",
+    "vite-plugin-svgr": "^2.2.2"
+  },
+  "optionalDependencies": {
+    "esbuild-linux-64": "^0.15.13"
```
Comment (dtfiedler) [inline package.json:41]: necessary?

File: src/components/inputs/Search/SearchBarHome/SearchBarHome.tsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { ReactComponent as SearchIcon } from '../../../icons/Search.svg';
+import './styles.css';
+import { searchBar } from '../../../../services/types';
+
+function SearchBarHome({ searchButtonAction }: searchBar) {
```
Comment (dtfiedler) [inline src/components/inputs/Search/SearchBarHome/SearchBarHome.tsx:5]:
```suggestion
function SearchBar({ searchButtonAction }: searchBar) {
```

we'll want to use the same search bar with some different effects in the nav bar, so let's give it a generic name

File: vite.config.ts [hunk 4]
```diff
@@ -1,7 +1,8 @@
 import { defineConfig } from 'vite';
 import react from '@vitejs/plugin-react';
+import svgr from 'vite-plugin-svgr';
 
 // https://vitejs.dev/config/
 export default defineConfig({
-  plugins: [react()],
+  plugins: [svgr(), react()],
```
Comment (dtfiedler) [inline vite.config.ts:7]: still needed with the mods on svg?

File: package.json (hunk 2 above)
Comment (atticusofsparta) [reply package.json:40]: Tried installing globally, didnt work, added it here. it SHOULD only install on linux 64 bit platforms from what I read.

File: src/components/inputs/Search/SearchBar/__tests__/SearchBar.test.tsx [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  test('render SearchBarHome', () => {
+    render(
+      <SearchBar
+        searchButtonAction={() => {
+          console.log('search');
+        }}
+      />,
+    );
+  });
+});
```
Comment (dtfiedler) [inline src/components/inputs/Search/SearchBar/__tests__/SearchBar.test.tsx:17]: `SearchBar` in describe and test

File: vite.config.ts (hunk 4 above)
Comment (atticusofsparta) [reply vite.config.ts:7]: yes, tried without, doesnt work

File: jest.config.json [hunk 6]
```diff
@@ -4,7 +4,8 @@
   "collectCoverage": false,
   "collectCoverageFrom": ["src/**/*.{ts,tsx,js,jsx}", "!src/**/*.d.ts"],
   "moduleNameMapper": {
-    "\\.(css|gif)$": "jest-transform-stub"
+    "\\.(css|gif)$": "jest-transform-stub",
+    "\\.svg": "<rootDir>/__mocks__/svg.js"
```
Comment (dtfiedler) [inline jest.config.json:8]: let's move to `src/__tests__/__mocks__/svg.js`

File: src/components/inputs/Search/SearchBar/__tests__/SearchBar.test.tsx [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+  test('render SearchBar', () => {
+    render(
+      <SearchBar
+        searchButtonAction={() => {
+          console.log('search');
+        }}
+      />,
+    );
+  });
```
Comment (dtfiedler) [inline src/components/inputs/Search/SearchBar/__tests__/SearchBar.test.tsx:15]:
```suggestion
  test('render SearchBar', () => {
   const stub = jest.fn();
    render(
      <SearchBar
        searchButtonAction={stub}
      />,
    );
  });
```

File: jest.config.json (hunk 6 above)
Comment (atticusofsparta) [reply jest.config.json:8]: Jest complains that svg.js isnt a test if we have mocks in there

File: src/components/inputs/Search/SearchBar/__tests__/SearchBar.test.tsx (hunk 5 above)
Comment (atticusofsparta) [reply src/components/inputs/Search/SearchBar/__tests__/SearchBar.test.tsx]: fixed

File: jest.config.json (hunk 6 above)
Comment (dtfiedler) [reply jest.config.json:8]: we can tell `jest` to ignore these 
```
  "moduleNameMapper": {
    "\\.(css|gif)$": "jest-transform-stub",
    "\\.svg": "<rootDir>/src/__tests__/__mocks__/svg.js"
  },
  "moduleDirectories": ["node_modules", "assets"],
  "testPathIgnorePatterns": ["src/__tests__/__mocks__"]
```

Comment (atticusofsparta) [reply jest.config.json]: Added ignore for mocks folder

File: custom.d.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+declare module '*.svg' {
+  import React = require('react');
+  export const ReactComponent: React.FC<React.SVGProps<SVGSVGElement>>;
+  const src: string;
+  export default src;
+}
```
Comment (dtfiedler) [inline custom.d.ts:6]: let's pull up `types.d.ts` from from `services/` and put these in there with a comment that these are necessary for Jest to handle `svgs`

File: src/components/inputs/Search/SearchBar/SearchBar.tsx [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    <div className="searchBar">
+      <input type="text" placeholder="sam" />
+      <button
+        className="searchButton"
+        onClick={() => {
+          searchButtonAction();
+        }}
+      >
+        <SearchIcon fill="black" stroke="white" width="18.51" />
+      </button>
```
Comment (dtfiedler) [inline src/components/inputs/Search/SearchBar/SearchBar.tsx:16]: per figma - the placeholder font size in the search bar looks heavy/bold and could use additional padding. also, the placeholder should likely be a prop passed into the component 
image
vs. 
image

Comment (dtfiedler) [reply src/components/inputs/Search/SearchBar/SearchBar.tsx:16]: as well is this in the console on load:
```
Error: <svg> attribute height: Expected length, "current".
setValueForProperty @ react-dom.development.js:798

```

File: src/types.d.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+export interface SmartweaveContractSource {
+  getContractState(
+    contractId: ArweaveTransactionId,
+  ): Promise<ArNSContractState | undefined>;
+}
+
+export type searchBar = {
+  searchButtonAction: any;
+};
```
Comment (dtfiedler) [inline src/types.d.ts:16]: duplicate file

File: src/components/inputs/Search/SearchBar/SearchBar.tsx [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import { ReactComponent as SearchIcon } from '../../../icons/Search.svg';
+import './styles.css';
+import { searchBar } from '../../../../types';
+
+function SearchBar({ searchButtonAction }: searchBar) {
```
Comment (dtfiedler) [inline src/components/inputs/Search/SearchBar/SearchBar.tsx:5]: let's also allow `headerText` and `footerText` to be optional props to pass in the text fields from the figma. for readability, I think it helps removing the destructring of props down into the function.

e.g.
```
function SearchBar(props: SearchBarProps){
    const { searchBarAction, headerText, footerText } = props;
    ...
}
```

File: src/types.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+// TODO: match this to a regex
+export type ArweaveTransactionId = string;
+
+export interface SmartweaveContractSource {
+  getContractState(
+    contractId: ArweaveTransactionId,
+  ): Promise<ArNSContractState | undefined>;
+}
+
+export type searchBar = {
```
Comment (dtfiedler) [inline src/types.ts:14]:
```suggestion
export type SearchBarProps = {
```

File: src/components/inputs/Search/SearchBar/SearchBar.tsx [hunk 13]
```diff
@@ -0,0 +1,43 @@
+import { ReactComponent as SearchIcon } from '../../../icons/Search.svg';
+import { ReactComponent as ArrowUpRight } from '../../../icons/ArrowUpRight.svg';
+import { useState, useEffect } from 'react';
+import './styles.css';
+import { SearchBarProps } from '../../../../types';
+
+function SearchBar(props: SearchBarProps) {
+  const {
+    buttonAction,
+    searchBarState,
```
Comment (dtfiedler) [inline src/components/inputs/Search/SearchBar/SearchBar.tsx:10]: this requires parent components to have context for various states of the search bar, which we don't necessarily want. instead, we can manage states internally so we don't have to reimplement what they are in parent components everytime this search bar is used.

TLDR, remove `searchBarState` and `onChangeHandler` from props, and we can add internal functions that mange those later on

File: src/components/inputs/Search/SearchBar/__tests__/SearchBar.test.tsx [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    const stub = jest.fn();
+    render(
+      <SearchBar
+        buttonAction={stub}
+        placeholderText="Enter a name"
+        searchState={"stub"}
+        searchBarState={"stub"}
+        onChangeHandler={stub}
+        headerText={''}
+        footerText={''}
```
Comment (dtfiedler) [inline src/components/inputs/Search/SearchBar/__tests__/SearchBar.test.tsx:17]:
```suggestion
        headerText={'Find a domain name'}
        footerText={'Names must be 1-32 characters. Dashes and underscores are permitted, but cannot be trailing characters and cannot be used in single character domains.'}
```
we can move these to constants.ts files once we have more

File: src/components/pages/Home/Home.tsx [hunk 15]
```diff
@@ -1,7 +1,36 @@
...
+  return (
+    <div className="home">
+      <SearchBar
+        buttonAction={searchAction}
+        placeholderText="Enter a name"
+        searchState={searchState}
+        searchBarState={searchBarState}
+        onChangeHandler={setSearchState}
+        headerText={''}
+        footerText={''}
```
Comment (dtfiedler) [inline src/components/pages/Home/Home.tsx:30]: per other comment, let's remove the state management from the parent and keep it internal to the component. we can add flags to indicate if the parent wants to see various states (e.g. `highlightError={error}`)

Comment (dtfiedler) [inline src/components/pages/Home/Home.tsx:30]: per other comment, let's remove the state management from the parent and keep it internal to the component. we can add flags to indicate if the parent wants to see various states (e.g. `highlightError={error}`)

```suggestion
        buttonAction={searchAction}
        placeholderText="Enter a name"
        headerText={'Find a domain name.'}
        footerText={'FIGMA TEXT HERE FOR CHARACTER DETAILS'}
```

File: src/types.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+export type SearchBarProps = {
+  buttonAction: any; // fn to run on search button click
+  searchBarState: string; // "search" "error" "success"
+  searchState: string; // value of the text input
+  onChangeHandler: any; // function to change state of input
+  placeholderText: string; //placeholder text of the input
+  headerText: any; // display text or component above search
+  footerText: any; // display text or component below search
```
Comment (dtfiedler) [inline src/types.ts:23]:
```suggestion
  buttonAction: () => void;
  placeholderText: string; 
  headerText: string; 
  footerText: string; 
```

**Review by dtfiedler — APPROVED** [review]
looks good, nice job
